fix: make flattenLast use object equality instead of identity

Christopher-Chianelli opened the pull request.

Christopher-Chianelli committed:
fix: make flattenLast use object equality instead of identity

triceo commented on core/src/test/java/ai/timefold/solver/core/impl/score/stream/common/uni/AbstractUniConstraintStreamTest.java:
```diff
@@ -2414,34 +2415,49 @@ public void expandToQuad() {
     @Override
     @TestTemplate
     public void flattenLastWithDuplicates() {
```
Why was this change to a test necessary?
I'd like to understand the extent to which we're actually potentially breaking people.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractFlattenLastNode.java:
```diff
@@ -169,4 +130,36 @@ public Propagator getPropagator() {
+
+        void retract(Consumer<OutTuple_> retractedConsumer) {
+            for (var i = newCount.intValue(); i < outTupleList.size(); i++) {
+                retractedConsumer.accept(outTupleList.get(i));
```
If the other consumers are called `insertConsumer` and `updateConsumer`, shouldn't this be called `retractConsumer`? The other two do not use past tense.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractFlattenLastNode.java:
```diff
@@ -41,117 +46,73 @@ public final void insert(InTuple_ tuple) {
     public final void update(InTuple_ tuple) {
-        List<OutTuple_> outTupleList = tuple.getStore(flattenLastStoreIndex);
-        if (outTupleList == null) {
+        Map<Object, FlattenLastMapEntry<FlattenedItem_, OutTuple_>> outTupleMap = tuple.getStore(flattenLastStoreIndex);
```
The full type of this map is crazy enough that I'd create a new interface for it as a shorthand.

triceo reviewed (commented):
Considering that we are making this change to improve the behavior of `precompute` and remove that one pesky exception from there, I am missing those changes in here - or are you planning another PR for that?

Also - technically, we are breaking backwards compatibility here. (Even though, in this case, acceptably.) We should write a note in the upgrade recipe, describing the change so that people know what to expect, and how to fix it.

Christopher-Chianelli committed:
chore: review comments

triceo committed:
Clarify `flattenLast` behavior and object equality

Clarified the behavior of `flattenLast` regarding object equality and its impact on tuple creation and retraction. Updated examples to demonstrate the correct implementation of `equals` and `hashCode` in relation to `flattenLast`.

triceo reviewed (approved):
Made small changes to the upgrade recipe.
LGTM.
